File: frontend/src/lib/reports.ts
import { apiFetch } from "./api";

export type ReportType = "daily" | "weekly" | "monthly";
export type ReleaseKind = "preview" | "on_demand" | "production";

export interface ReportRun {
  id: string;
  journey_id: string | null;
  report_type: ReportType;
  kind: ReleaseKind;
  release_state: string;
  idempotency_key: string;
  period_start: string | null;
  period_end: string | null;
  cutoff_at: string | null;
  timezone: string;
  generation_version: string;
  is_partial_period: boolean;
  failure_reason: string | null;
  generated_at: string | null;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  snapshot: Record<string, any>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  detail: Record<string, any>;
  created_at: string;
}

export interface ReportPreference {
  report_type: ReportType;
  enabled: boolean;
  local_time: string;
  weekdays: number[];
  day_of_week: number | null;
  day_of_month: number | null;
  timezone: string;
  requested_channel: string;
  effective_channel: string;
}

export interface ReportPreferences {
  daily: ReportPreference;
  weekly: ReportPreference;
  monthly: ReportPreference;
}

export const reportApi = {
  generate: (journeyId: string, reportType: ReportType, releaseKind: ReleaseKind = "preview") =>
    apiFetch<ReportRun>(`/journeys/${journeyId}/reports/${reportType}/generate`, {
      method: "POST",
      body: JSON.stringify({ release_kind: releaseKind }),
    }),
  list: (journeyId: string, reportType?: ReportType) =>
    apiFetch<ReportRun[]>(`/journeys/${journeyId}/reports${reportType ? `?report_type=${reportType}` : ""}`),
  get: (journeyId: string, reportRunId: string) => apiFetch<ReportRun>(`/journeys/${journeyId}/reports/${reportRunId}`),
  preferences: () => apiFetch<ReportPreferences>("/reports/preferences"),
  updatePreference: (
    reportType: ReportType,
    body: { enabled: boolean; local_time: string; weekdays: number[]; day_of_week: number | null; day_of_month: number | null; timezone: string; requested_channel: string },
  ) => apiFetch<ReportPreference>(`/reports/preferences/${reportType}`, { method: "PUT", body: JSON.stringify(body) }),
};

File: backend/app/schemas/reports.py
from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class Strict(BaseModel):
    model_config = ConfigDict(extra="forbid")


ReportType = Literal["daily", "weekly", "monthly"]
ReleaseKind = Literal["preview", "on_demand", "production"]


class ReportGenerateRequest(Strict):
    release_kind: ReleaseKind = "preview"
    year: int | None = Field(default=None, ge=2020, le=2100)
    month: int | None = Field(default=None, ge=1, le=12)
    week_start: str | None = Field(default=None, max_length=10)


class ReportRunOut(Strict):
    id: UUID
    journey_id: UUID | None
    report_type: ReportType
    kind: ReleaseKind
    release_state: str
    idempotency_key: str
    period_start: datetime | None
    period_end: datetime | None
    cutoff_at: datetime | None
    timezone: str
    generation_version: str
    is_partial_period: bool
    failure_reason: str | None
    generated_at: datetime | None
    ready_to_send_at: datetime | None
    snapshot: dict[str, Any]
    detail: dict[str, Any]
    created_at: datetime


class ReportPreferenceOut(Strict):
    report_type: ReportType
    enabled: bool
    local_time: str
    weekdays: list[int]
    day_of_week: int | None
    day_of_month: int | None
    timezone: str
    requested_channel: str
    effective_channel: str


class ReportPreferencesOut(Strict):
    daily: ReportPreferenceOut
    weekly: ReportPreferenceOut
    monthly: ReportPreferenceOut


class ReportPreferenceUpdate(Strict):
    enabled: bool
    local_time: str = Field(pattern=r"^([01]\d|2[0-3]):[0-5]\d$")
    weekdays: list[int] = Field(default_factory=list)
    day_of_week: int | None = Field(default=None, ge=0, le=6)
    day_of_month: int | None = Field(default=None, ge=1, le=31)
    timezone: str = Field(min_length=1, max_length=64)
    requested_channel: Literal["off", "in_app", "email", "both"] = "in_app"
